### PR TITLE
Implement scala http4s server generator

### LINKED ISSUE #17428: [REQ] Create code gen for scala http4s library
### Is your feature request related to a problem? Please describe.

There is lack of server codegen for pretty popular scala http4s framework, it would be good to have one.

## Describe the solution you'd like

Codegen targeted to modern http4s stack with circe as json library.

## Describe alternatives you've considered

There is guardrail project with needed codegen, but guardrail project feels like to be abandoned, and it's current codeden lacks of needed possibilities. Also codebase of guardrail is pretty complex.

## Additional context

Actually I would like to provide MR with http4s codegen with feature request.

### PR DESCRIPTION
**Pull request by mikkka**
### PR checklist
 
- [ x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x ] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [ x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [ x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming 7.1.0 minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [ x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

PR closes #17428 

@fish86, please have a look

### CODE REVIEW

**Comment by mikkka** [comment]
@clasnake @jimschubert @shijinkui @ramzimaalej @chameleon82 @Bouillie 
Please, have a look.

**Comment by wing328** [comment]
Thanks for the PR but your commit (as shown in the Commits tab) is not linked to your Github account, which means this PR won't count as your contribution in https://github.com/OpenAPITools/openapi-generator/graphs/contributors.

Let me know if you need help fixing it.

Ref: https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-can-i-update-commits-that-are-not-linked-to-my-github-account

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/ScalaHttp4sServerCodegen.java [hunk 1]
```diff
@@ -0,0 +1,796 @@
+/*
+ * Copyright 2018 OpenAPI-Generator Contributors (https://openapi-generator.tech)
+ * Copyright 2018 SmartBear Software
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/ScalaHttp4sServerCodegen.java:3]: for new generator, I don't think we need this line.

**Comment by wing328** [comment]
what about also auto-generating a build.sbt file as well so as to build the output to ensure there's no syntax error?

**Comment by mikkka** [comment]
@wing328 
I fixed

**Comment by a-khakimov** [comment]
Shouldn't this implementation be included in the list of generators in the docs? https://openapi-generator.tech/docs/generators

**Comment by wing328** [comment]
I ran `java -jar modules/openapi-generator-cli/target/openapi-generator-cli.jar generate -g scala-http4s-server -i modules/openapi-generator/src/test/resources/3_0/petstore.yaml -o /tmp/scala/` and then `sbt test` in the output but got the following errors:
```
[error] /private/tmp/scala/src/main/scala/org/openapitools/apis/UserApi.scala:287:102: ']' expected but ';' found.
[error]       username: Refined[String, MatchesRegex[&quot;^[a-zA-Z0-9]+[a-zA-Z0-9\\.\\-_]*[a-zA-Z0-9]+$&quot;]],
[error]                                                                                                      ^
[error] /private/tmp/scala/src/main/scala/org/openapitools/apis/UserApi.scala:287:64: type application is not allowed for postfix operators
[error]       username: Refined[String, MatchesRegex[&quot;^[a-zA-Z0-9]+[a-zA-Z0-9\\.\\-_]*[a-zA-Z0-9]+$&quot;]],
[error]                                                                ^
[error] /private/tmp/scala/src/main/scala/org/openapitools/apis/UserApi.scala:287:103: illegal start of definition
[error]       username: Refined[String, MatchesRegex[&quot;^[a-zA-Z0-9]+[a-zA-Z0-9\\.\\-_]*[a-zA-Z0-9]+$&quot;]],
```

can you please take a look when you've time?

**Comment by wing328** [comment]
can you please create new config file for your new scala server generator based on ./bin/configs/scala-pekko-http-server.yaml  ?

and then update .github/workflows/samples-scala.yaml accordingly so that the CI will test it moving forward.

**Comment by wing328** [comment]
@a-khakimov we will update the doc later after merging the PR

**Comment by mikkka** [comment]
Fixed errors, added configs.

**Comment by mikkka** [comment]
@wing328 please have a look

**Review by a-khakimov — APPROVED** [review]

File: bin/configs/scala-http4s-server.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+generatorName: scala-http4s-server
+outputDir: samples/server/petstore/scala-http4s-server
```
Comment (wing328) [inline bin/configs/scala-http4s-server.yaml:2]: can you please commit the new files for scala-http4s-server petstore as well?

then CI will run the tests on the new sample.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/ScalaHttp4sServerCodegen.java [hunk 3]
```diff
@@ -0,0 +1,851 @@
...
+import org.openapitools.codegen.model.*;
+import org.openapitools.codegen.utils.ModelUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class ScalaHttp4sServerCodegen extends DefaultCodegen implements CodegenConfig {
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/ScalaHttp4sServerCodegen.java:30]: have you considered extending `modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractScalaCodegen.java` instead?

Comment (mikkka) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/ScalaHttp4sServerCodegen.java:30]: Done

File: bin/configs/scala-http4s-server.yaml (hunk 2 above)
Comment (mikkka) [reply bin/configs/scala-http4s-server.yaml:2]: Done

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/ScalaHttp4sServerCodegen.java (hunk 3 above)
Comment (wing328) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/ScalaHttp4sServerCodegen.java:30]: Understood. It's not a must as extending DefaultCodegen is also fine.

**Comment by wing328** [comment]
looks good. let's give it a try

**Comment by wing328** [comment]
samples, docs updated via 9b7528d8c8c

**Comment by mikkka** [comment]
@wing328 thank you for support, when it will be available in distro of openapi generator (and sbt plugin for codegen)?

**Comment by wing328** [comment]
likely we will cut a release by the end of this month

**Pull request merged**